Fix reading internally compressed files + Update Parso + Other Stuff

thesuperzapper opened the pull request:
This PR completely rewrites the splitting logic, and should allow reading of internally compressed SAS files resolving #38, **but this should be confirmed by others experimenting with their particular files.** (In doing this, we bump the Parso version to 2.0.10.)

Additionally, we resolve #43 through the addition of `forceLowerCaseColumnNames` as a parameter. This defaults to false for backwards compatibility. 
E.g (PySpark):
```df = sqlContext.read.format("com.github.saurfang.sas.spark").load("cars.sas7bdat", forceLowerCaseColumnNames=True)```

Further, we expose `minPartitions` as a parameter intended to define the minimum number of splits for the SAS file. However, without a bit of further rewriting, it’s only a hint, and will not necessarily be honoured. 
E.g (PySpark):
```df = sqlContext.read.format("com.github.saurfang.sas.spark").load("cars.sas7bdat", minPartitions=5)```

Sorry for mixing multiple changes in one PR 😊.

kylebarron commented:
Sorry for the newbie question, but how do you build this? I'd like to test this on some data that was previously not working. I tried to build with

```
git clone https://github.com/thesuperzapper/spark-sas7bdat
cd spark-sas7bdat
sbt .
```
and it gave me:
```
> sbt .
[info] Loading project definition from /disk/homedirs/barronk-dua51929/spark-sas7bdat/project
[info] Set current project to spark-sas7bdat (in build file:/disk/homedirs/barronk-dua51929/spark-sas7bdat/)
[error] Expected letter
[error] Expected symbol
[error] Expected '!'
[error] Expected '+'
[error] Expected '++'
[error] Expected '^'
[error] Expected '^^'
[error] Expected 'debug'
[error] Expected 'info'
[error] Expected 'warn'
[error] Expected 'error'
[error] Expected ';'
[error] Expected end of input.
[error] Expected 'early('
[error] Expected '-'
[error] Expected '--'
[error] Expected 'show'
[error] Expected 'all'
[error] Expected '*'
[error] Expected '{'
[error] Expected project ID
[error] Expected configuration
[error] Expected key
[error] .
[error] ^
```

I'm using `sbt` version 2.0.0 and I set `SPARK_HOME` to an installation of Spark 2.2.

kylebarron commented:
This does work for me on at least one dataset where I had previously (with the most recently released jar) gotten a `java.lang.reflect.InvocationTargetException` similar to the [original issue](https://github.com/saurfang/spark-sas7bdat/issues/32#issue-300675870).

kylebarron commented:
It really would be great to include a test using a file that previously didn't work. I think someone on the other issue posted such a dataset?

thesuperzapper commented:
@kylebarron 

The cause of the error was related to incorrect page boundary calculations, and the general misreading of sas7bdat files with PAGE_META_TYPE_2 pages. (The name given by the parso project)

The old version reads this file [recapture_test_compressed.zip](https://github.com/saurfang/spark-sas7bdat/files/2579038/recapture_test_compressed.zip), as having 90536 rows, when it has 98366. This can be used to test PAGE_META_TYPE_2 containing files. (And its quite small at 13MB)

The actual crash error, was caused by really unlucky splits on files containing PAGE_META_TYPE_2, meaning recreating it will be challenging, as splits are calculated in a system-dependant way.

The [file provided by vivard](https://github.com/saurfang/spark-sas7bdat/issues/32#issuecomment-412535049), is quite large uncompressed (270MB) so including it in a test might be bad idea, but its much more likely to induce a split as a result of its size, and on most systems will actually crash.

@saurfang which ones do you want me to include?

kylebarron commented:
Yeah, I agree that it wouldn't be good to include a file that large in the git repository. I don't have much knowledge of the SAS format internals or of Scala, so I don't really have anything else to add.

Tagar commented:
>> The actual crash error, was caused by really unlucky splits on files containing PAGE_META_TYPE_2, 
>> meaning recreating it will be challenging, as splits are calculated in a system-dependant way.

If input sas file for Spark is in HDFS, 
`mapreduce.input.fileinputformat.split.minsize`,  `mapreduce.input.fileinputformat.split.maxsize` 
should be able to control split sizes. 

For smaller files we can drive `mapreduce.input.fileinputformat.split.maxsize`  way down to 
make a bigger number of splits. Same file can be retested multiple times with a different set of 
split sizes, to try to trigger different issues.

PySpark example:
```python
sc = ...  # type: SparkContext

sc._jsc.hadoopConfiguration().setInt("mapred.min.split.size", min_value)
sc._jsc.hadoopConfiguration().setInt("mapred.max.split.size", max_value)

```

`mapred.max.split.size` is an older name for `mapreduce.input.fileinputformat.split.maxsize` (the former now is deprecated in newer versions of Hadoop, but might still work too).

Tagar commented on build.sbt:
```diff
@@ -5,11 +5,11 @@ organization := "com.github.saurfang"
 crossScalaVersions := Seq("2.10.7", "2.11.12")
 
-scalacOptions ++= Seq("-target:jvm-1.7" )
+scalacOptions ++= Seq("-target:jvm-1.7")
```
minor: Spark 2.2 and higher require Java 8.. not sure if it's important here.

Tagar commented on src/main/scala/com/github/saurfang/sas/spark/DefaultSource.scala:
```diff
@@ -47,6 +47,24 @@ class DefaultSource
-    SasRelation(checkPath(parameters), schema)(sqlContext)
+    val path = checkPath(parameters)
+    
+    val forceLowerCaseColumnNames = parameters.getOrElse("forceLowerCaseColumnNames", "false")
```
minor: should there be an option `forceUpperCaseColumnNames` too?

thesuperzapper replied:
I dont think we need this, as is pretty bad spark practice?

Tagar commented on src/main/scala/com/github/saurfang/sas/mapred/SasRecordReader.scala:
```diff
@@ -27,131 +27,157 @@ import org.apache.hadoop.mapred.{FileSplit, InputSplit, RecordReader}
+  private val codec = new CompressionCodecFactory(job).getCodec(filePath)
   if (codec != null) {
-    throw new IOException("SASRecordReader does not support reading compressed files")
+    throw new IOException("SASRecordReader does not support reading compressed files.")
```
This PR reads "Fix reading internally compressed files.. " - so SAS-compressed files are still not supported?

thesuperzapper replied:
This is referring to files that are zipped on the file system level.

Tagar commented on src/main/scala/com/github/saurfang/sas/mapred/SasRecordReader.scala:
```diff
@@ -27,131 +27,157 @@ import org.apache.hadoop.mapred.{FileSplit, InputSplit, RecordReader}
+  
+  //// Initialize
+  // Variables
+  private var recordCount: Int = 0
```
Is this records per split? or per file.. if it's per file or if file is not splittable, `long` datatype might be better IMO

> "The maximum size of a SAS data set in a Direct Access Bound Library is limited by the maximum size of the library, which is about 2986 GB on 3390 volumes."
> 
> http://support.sas.com/kb/8/213.html
> 
>

thesuperzapper replied:
Agreed

kraj007 commented:
Tested on relatively small compressed  datasets upto 1 GB , results looks promising.  I have few datasets upto 15 gb.
I will update with results. Thanks @thesuperzapper  for your contribution !!

Tagar commented:
I have this file handy that we could add to tests too.
https://github.com/Tagar/dropbox/blob/master/10rec.sas7bdat
I know it was broken in the past too https://github.com/BioStatMatt/sas7bdat/issues/12#issuecomment-339433779
Also it's a very wide dataset (14,461 columns but just 10 rows) which might be a good idea for tests to cover such datasets too. 
Thanks.

saurfang commented:
Since the test resources are not included in the published JAR, I think it'd be fine to include test dataset <50MB (and considering we git will compress it further.)

Also, current test suites already actively test splitting logic by setting `BLOCK_SIZE` like so https://github.com/saurfang/spark-sas7bdat/blob/master/src/test/scala/com/github/saurfang/sas/SasInputFormatSpec.scala#L13 Including a new dataset that is previously broken would be ideal.
See this test that compares `sas7bdat` with reference `csv` result https://github.com/saurfang/spark-sas7bdat/blob/master/src/test/scala/com/github/saurfang/sas/SasInputFormatSpec.scala#L13

Thanks everyone for the great energy on this issue.

thesuperzapper commented:
I have made the following changes:

**General Improvements:**
* schema parsing logic
* error handling for incorrect schema declarations
* updated licence headers

**New Method: **
* `readFormattedColsAsDecimal` which casts formatted numeric SAS columns as decimals with the same precision/width. (Think currency, which might be represented as a COMMA22.2, would become a DecimalType(22,2))

**New test cases:**
* Ensure that a more complex (multi typed) SAS file is read correctly.
* Ensure that internally compressed files are read the same as ones that aren’t.
* Ensure that `readFormattedColsAsDecimal` behaves correctly.

thesuperzapper commented:
Here is a pre-built version for Spark 2.2, with the above changes. [spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v2.zip](https://github.com/saurfang/spark-sas7bdat/files/2584939/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v2.zip)

PhoenixDai commented:
tested with Spark 2.3.0, works well too.

thesuperzapper commented:
I am currently working on improving the schema logic, and making the test cases less heavy (And based off CSV rather than Parquet)

thesuperzapper commented:
Here is the latest version, should be almost final: [spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v3.zip](https://github.com/saurfang/spark-sas7bdat/files/2596976/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v3.zip)

The current list of parameters is:
* extractLabel (Default: False)
   * Boolean: extract column labels as column comments.
* forceLowercaseNames (Default: False)
    * Boolean: if schema inference will force column names to lower case.
* inferDecimal (Default: False)
   * Boolean: if inference will automatically infer numeric columns (With sas format width > 0 and sas format precision >0), as DecimalType(Width, Precision)
* inferDecimalScale (Default: False)
   * Int: The scale of inferred decimals. (Defaults: each column's format width)
* inferFloat (Default: False)
    * Boolean: if inference will automatically infer numeric columns (With <=4 bytes), as Floats.
* inferInt (Default: False)
    * Boolean: if inference will automatically infer numeric columns (With <=4 bytes, format width >0 and format precision = 0), as Integers.
* inferLong (Default: False)
    * Boolean: if inference will automatically infer numeric columns (With <=8 bytes, format width >0 and format precision = 0), as Longs.
* inferShort (Default: False)
    * Boolean: if inference will automatically infer numeric columns (With <=2 bytes, format width >0 and format precision = 0), as Longs.
* minPartitions (Default: 0)
    * Int: The minimum number splits to perform. (Adherence is not guaranteed)

kraj007 commented:
> Here is the latest version, should be almost final: [spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v3.zip](https://github.com/saurfang/spark-sas7bdat/files/2596976/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v3.zip)
> 
> The current list of parameters is:
> 
> * extractLabel (Default: False)
>   
>   * Boolean: extract column labels as column comments.
> * forceLowercaseNames (Default: False)
>   
>   * Boolean: if schema inference will force column names to lower case.
> * inferDecimal (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With sas format width > 0 and sas format precision >0), as DecimalType(Width, Precision)
> * inferDecimalScale (Default: False)
>   
>   * Int: The scale of inferred decimals. (Defaults: each column's format width)
> * inferFloat (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=4 bytes), as Floats.
> * inferInt (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=4 bytes, format width >0 and format precision = 0), as Integers.
> * inferLong (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=8 bytes, format width >0 and format precision = 0), as Longs.
> * inferShort (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=2 bytes, format width >0 and format precision = 0), as Longs.
> * minPartitions (Default: 0)
>   
>   * Int: The minimum number splits to perform. (Adherence is not guaranteed)

getting error as : 
xception in thread "main" java.lang.NoSuchMethodError: com.github.saurfang.sas.spark.package$SasContext.sasFile$default$2()Lorg/apache/hadoop/mapred/JobConf;

with attached jar.

i am reading sas files as :
 val sas_df=spark.sqlContext.sasFile(args(0)).cache()

thesuperzapper commented:
@saurfang 

I am mostly finished here, the only tasks remaining are documentation. (Do you want me to do this?)

I would like to highlight that there is one small breaking change, which is that you can no longer specify jobConf directly using `sqlContext.sasFile(path, conf)`, and instead must modify the sparkContext's hadoopConf e.g.

````
val conf = sc.hadoopConfiguration
conf.setInt("fs.local.block.size", BLOCK_SIZE)
conf.setInt("mapred.min.split.size", BLOCK_SIZE)
```

Tagar commented:
We also have confirmation from our users that internally compressed sas7bdat files can now be read correctly. At this time, they didn't find older files that are known were not working in previous versions, but ones they have tried were working correctly. Thank you for these improvements.

kraj007 commented:
> sqlContext.sasFile(path, conf)

@thesuperzapper Could you please update documentation for scala API.
I tried with spark.sqlContext.sasFile and was getting error as : Exception in thread "main" java.lang.NoSuchMethodError. for sqlContext.sasFile

thesuperzapper commented:
@kraj007 what exactly did you write?

The docs on the readme should still work, please tell me if they don't.

E.g.
```
import com.github.saurfang.sas.spark._
val random = sqlContext.sasFile("src/test/resources/random.sas7bdat")
```

kraj007 commented:
> @kraj007 what exactly did you write?
> 
> The docs on the readme should still work, please tell me if they don't.
> 
> E.g.
> 
> ```
> import com.github.saurfang.sas.spark._
> val random = sqlContext.sasFile("src/test/resources/random.sas7bdat")
> ```
trying with below , it works fine with first jar attached but issues with latest jar.


import com.github.saurfang.sas.spark._

 val spark = SparkSession.builder().appName("Sas_Spark").getOrCreate()
 
 val df=spark.sqlContext.sasFile("tmp/cars.sas7bdat").cache()





Exception in thread "main" java.lang.NoSuchMethodError: com.github.saurfang.sas.spark.package$SasContext.sasFile$default$2()Lorg/apache/hadoop/mapred/JobConf;

thesuperzapper commented:
@kraj007 what version of spark are you using?

kraj007 commented:
> @kraj007 what version of spark are you using?

spark : version 2.1.1
Scala version 2.11.8 
parso-2.0.10.jar
spark-sas7bdat_2.11-2.1.0_SNAPSHOT.jar  as external dependecies 

I am using IntelliJ

thesuperzapper commented:
@kraj007 I know the issue, and will fix it tomorrow. (Midnight in NZ right now)

Tagar commented:
Just out of curiosity why all of them default to `False`?
Current setup makes all numeric fields inferred as Double.

I think it would be nice to change to defaults for `extractLabel`, `inferFloat`, `inferInt`, `inferLong`, `inferShort` to **True**. It would create more compact output presentations for larger sas7bdat files. 

Also Double is not good as a catch-all numeric datatype for all use cases. For example, some long ids will be unusable due to lost precision in Double datatype - Long will be better for those. 

Would be nice to have this in documentation too.

Thanks!!


> The current list of parameters is:
> 
> * `extractLabel` (Default: False)
>   
>   * Boolean: extract column labels as column comments.
> * `forceLowercaseNames` (Default: False)
>   
>   * Boolean: if schema inference will force column names to lower case.
> * `inferDecimal` (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With sas format width > 0 and sas format precision >0), as DecimalType(Width, Precision)
> * `inferDecimalScale` (Default: each column's format width)
>   
>   * Int: The scale of inferred decimals.
> * `inferFloat` (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=4 bytes), as Floats.
> * `inferInt` (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=4 bytes, format width >0 and format precision = 0), as Integers.
> * `inferLong` (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=8 bytes, format width >0 and format precision = 0), as Longs.
> * `inferShort` (Default: False)
>   
>   * Boolean: if inference will automatically infer numeric columns (With <=2 bytes, format width >0 and format precision = 0), as Longs.
> * `minPartitions` (Default: 0)
>   
>   * Int: The minimum number splits to perform. (Adherence is not guaranteed)

thesuperzapper commented:
@Tagar the reason for defaulting to Double, is because SAS doesn't have a concept of Longs, instead it just rounds the numeric types on disk when the user looks at it (column formatters). To infer if a column is a long, I check the column formatter, and if its precision is 0, meaning I might reduce precision if someone hasn't defined their formatters correctly.

(Also, making this change would be inconsistent with previous versions of this reader)

kraj007 commented:
> @kraj007 I know the issue, and will fix it tomorrow. (Midnight in NZ right now)



> @kraj007 I know the issue, and will fix it tomorrow. (Midnight in NZ right now)

Sure @thesuperzapper  , would like to test datasets with latest changes again on spark 2.X

thesuperzapper commented:
@kraj007, can you check this version?
[spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v4.zip](https://github.com/saurfang/spark-sas7bdat/files/2607224/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v4.zip)

If it doesn't work, can you send Spark executor logs?

thesuperzapper commented:
@kraj007

I bet your args(0) is a JobConf object (is it?), If you need to pass configs with the new version, set them using `SparkContext.hadoopConfiguration`.

kraj007 commented:
> @kraj007
> 
> I bet your args(0) is a JobConf object (is it?), If you need to pass configs with the new version, set them using `SparkContext.hadoopConfiguration`.

i ran using 
val sas_df=spark.sqlContext.sasFile(args(0),JobConf).cache()

args(0) is filename , still got same error.

Howerever checked the package.class files in working first version jar as 
public Dataset<Row> sasFile(String filePath, JobConf conf)

for latest jar its 
public Dataset<Row> sasFile(String filePath).

thesuperzapper commented:
@kraj007 can you try hard coding the string for args(0), and see if it works then?

kraj007 commented:
> @kraj007 can you try hard coding the string for args(0), and see if it works then?

Tried with hardcoding the file_path in code , buit facing the same issues with latest jar attached above whereas it works fine with spark-sas7bdat_2.11-2.1.0_SNAPSHOT.zip

thesuperzapper commented:
@kraj007
Here is a version compiled for Spark 2.1: [spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v4_SPARK-2.1.zip](https://github.com/saurfang/spark-sas7bdat/files/2614814/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v4_SPARK-2.1.zip)

Realistically, you should use the default source reader rather than the implicit reader anyway, as SQLContext is deprecated in Spark 2.0+

E.g.
```
df = sc.read.format("com.github.saurfang.sas.spark").load("cars.sas7bdat")
```

thesuperzapper commented:
I have fixed some bugs with reading corrupt/invalid SAS files, which used to hang if you tried to read them, if you point the reader at a non-sas file, it should throw an error now, as it check for >0 columns. 

We now have the following new parameter:
* `metadataTimeout` (Default: 60)
   * Int: the number of seconds to allow reading of file metadata, before throwing an error.

thesuperzapper commented:
@saurfang 
I have now written the docs and release notes.

thesuperzapper commented:
Here are the jars for this version.

[spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v5.zip](https://github.com/saurfang/spark-sas7bdat/files/2622194/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v5.zip)

Tagar commented on README.md:
```diff
@@ -1,93 +1,100 @@
+  * Int: number of seconds to allow reading of file metadata _(stops corrupt files hanging)_
+* `minPartitions` _(Default: 0)_
+  * Int: minimum number splits to use when reading
+  
```
Would it be also possible to add comment you left earlier at https://github.com/saurfang/spark-sas7bdat/pull/44#issuecomment-440840807 - this wasn't obvious for me. thank you.

> the reason for defaulting to Double, is because SAS doesn't have a concept of Longs, instead it just rounds the numeric types on disk when the user looks at it (column formatters). To infer if a column is a long, I check the column formatter, and if its precision is 0, meaning I might reduce precision if someone hasn't defined their formatters correctly.
> 
> (Also, making this change would be inconsistent with previous versions of this reader)
> 
>

thesuperzapper replied:
Done

thesuperzapper commented:
__Here is the new version:__
[spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v6.zip](https://github.com/saurfang/spark-sas7bdat/files/2624594/spark-sas7bdat_2.11-2.1.0_SNAPSHOT_v6.zip)

__Added new paramaters:__
* `minSplitSize` _(Default: `mapred.min.split.size`)_
  * Long: minimum byte length of input splits _(splits are always at least 1MB, to ensure correct reads)_
* `maxSplitSize` _(Default: `mapred.max.split.size`)_
  * Long: maximum byte length of input splits, _(can be decreased to force higher parallelism)_

PhoenixDai commented:
Hi @kraj007 , I tested the following settings with the most recent version on Spark 2.4.0. All working well. Please feel free to share your code snippets if you'd like me to help debugging. 

val sqlContext: SQLContext = new SQLContext(sc)
val df: DataFrame = sqlContext.sasFile(args(0))

val spark = SparkSession.builder().getOrCreate()
val df = spark.read.format("com.github.saurfang.sas.spark").load(args(0))
val df2 = spark.sqlContext.sasFile(args(0))

Tagar commented on src/main/scala/com/github/saurfang/sas/spark/SasRelation.scala:
```diff
@@ -16,119 +16,251 @@
+
+    // Set min/max split sizes if provided.
+    if (!minSplitSize.isEmpty) {conf.setLong("mapred.min.split.size", minSplitSize.get)}
+    if (!maxSplitSize.isEmpty) {conf.setLong("mapred.max.split.size", maxSplitSize.get)}
```
`mapred.min.split.size` is deprecated in favor of `mapreduce.input.fileinputformat.split.minsize`.

https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/DeprecatedProperties.html

https://hadoop.apache.org/docs/r2.6.2/hadoop-project-dist/hadoop-common/DeprecatedProperties.html

Same for `mapred.max.split.size`.

saurfang replied:
I don't think we should expose these two as parameters if all we do is to pass them down as Spark configs. It is fine to educate users about these configurations in the readme. AFAIK no other data source in Spark have these kind of proxy configs.

Tagar replied:
my 1.5 cents - Good point. On the other hand, `spark-sas7bdat` is tightly coupled with hadoop already .. I agree, that moving this just to docs makes sense, as there is a way to set same configs through spark config. 

ps. Not every Spark user runs Spark on Hadoop.. might be a good idea to refactor it in the future to make `spark-sas7bdat` more abstract so somebody could use it for Spark-on-Kubernetes, for example. I think the splitting logic would be the same for `S3` storage for example?

thesuperzapper replied:
@Tagar I am not 100% sure, but the current code should support S3 URI's already? As spark uses hadoop code for reading anyway, even if running on K8S.

@saurfang, it's up to you, but the reason I include them here, is because people might want to specify max/min size at the file level, rather than globally for their spark context. (Notice that I make a copy of the `sparkContext.hadoopConfiguration` before adding these configs, rather than changing the users global conf)

saurfang commented on README.md:
````diff
@@ -1,93 +1,104 @@
-USING com.github.saurfang.sas.spark
-OPTIONS (path "cars.sas7bdat")
-```
+__NOTE:__ we currently only support reading
````
did you mean to say something here?

thesuperzapper replied:
I mean to say we don't current sport writing sas7bdat, but I can make that more clear.

thesuperzapper commented on src/main/scala/com/github/saurfang/sas/spark/package.scala:
```diff
@@ -16,20 +16,16 @@
+      val sasRelation = SasRelation(location = filePath)(sqlContext)
       sqlContext.baseRelationToDataFrame(sasRelation)
     }
   }
```
As the SQLContext is so core to the V1 DataSource API, we might as well keep it as the preferred thing, until we create a new package for the V2 API, which was released in Spark 2.3.

thesuperzapper replied:
However, I have added an implicit for `DataFrameReader`

thesuperzapper commented:
@saurfang as long as you are happy with the readme/news, feel free to merge.

When you release the jar, don't forget to change `2.1.0_SNAPSHOT` to `2.1.0` in build.sbt
